openfisca_uk/calibration/reweighting.py

```diff
@@ -0,0 +1,188 @@
+from pathlib import Path
+import h5py
+from tqdm import tqdm
+from openfisca_uk import Microsimulation
+import pandas as pd
+import numpy as np
+import tensorflow as tf
+import logging
+
+tf.get_logger().setLevel("INFO")
+logging.basicConfig(level=logging.INFO)
+
+sim = Microsimulation(adjust_weights=False)
+parameters = sim.simulation.tax_benefit_system.parameters
+statistics = parameters.statistics
+
+START_YEAR = 2019
+END_YEAR = 2022
+
+FORCE_ALL_METRIC_IMPROVEMENT = False  # Don't save weights unless all (program, year, aggregate/caseload) tests improve. If false, regressions will be logged.
+MODIFICATION_PENALTY = 1e-11
+
+household_population = sim.calc("people", map_to="household").values
+household_region = sim.calc("region").values
+regions = list(pd.Series(household_region).unique())
+survey_num_households = len(sim.calc("household_id"))
+
+
+def loss(
+    weight_modification: np.array, include_modification_penalty: float = True
+) -> float:
+    # Ensure the weight modification doesn't increase the total number of UK households
+    weight_modification = (
+        weight_modification
+        - tf.reduce_sum(weight_modification, axis=0) / survey_num_households
+    )
+    l = 0
+    for year in range(START_YEAR, END_YEAR + 1):
+        instant_str = f"{year}-01-01"
+        statistic_set = statistics(instant_str)
+        variables = list(statistic_set.aggregate._children)
+        weights = sim.calc("household_weight", period=year).values
+        modified_weights = tf.nn.relu(
+            weight_modification[year - START_YEAR] + weights
+        )
+        for variable in variables:
+            if variable in statistic_set.aggregate._children:
+                values = sim.calc(
+                    variable, period=year, map_to="household"
+                ).values
+                # Calculate aggregate error
+                agg = tf.reduce_sum(modified_weights * values)
+                target = getattr(statistic_set.aggregate, variable)
+                l += ((agg - target) / 1e9) ** 2
```

Comment:
Could you move this 1e9 into a named constant like `MODIFICATION_PENALTY` (also the 1e6 for caseload)

Alternatively, we could normalize each deviation by dividing by `target` and then specify an importance vector for each variable (possibly set at the current aggregates with a 1000x boost for participation to mimic the current logic).